### PR TITLE
Feature/diagnostics

### PR DESCRIPTION
**Pull request by rattuscz**
Still a bit of work in progress:
 - what should be in the debug panel? right now it's just a list of dispatched events plus list of registered listeners
 - Also would love better icon if someone could find/create :-)

as mentioned it adds two new dependencies - Tracy and psr/log. I've kept them in require-dev but psr/log should probably go to normal require as it does not need to be only for development.

### CODE REVIEW

File: src/DI/EventDispatcherExtension.php [hunk 1]
```diff
@@ -45,6 +50,33 @@ public function loadConfiguration(): void
 			$eventDispatcherDefinition
 				->setFactory(EventDispatcher::class);
 		}
+
+		if (!$this->config->debug && $this->config->logger === null) {
+			return;
+		}
+
+		$wrapped = $builder->addDefinition($this->prefix('wrappedOriginal'))
```
Comment (f3l1x) [inline src/DI/EventDispatcherExtension.php]: I'd prefer sth like that:

```
if (debug) {
  // register DiagnosticDispatcher ( normal Dispatcher )
} else {
  // register normal EventDispatcher
}
```

File: src/Diagnostics/DiagnosticDispatcher.php [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+	/** @var LoggerInterface|null */
+	private $logger = null;
+
+	public function __construct(EventDispatcherInterface $original)
+	{
+		$this->original = $original;
+	}
+
+	public function setPanel(Panel $panel): void
```
Comment (f3l1x) [inline src/Diagnostics/DiagnosticDispatcher.php]: What it has connection to panel?

File: src/Diagnostics/DiagnosticDispatcher.php [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+	{
+		$this->panel = $panel;
+		$panel->setDispatcher($this);
+	}
+
+	public function setLogger(LoggerInterface $logger): void
```
Comment (f3l1x) [inline src/Diagnostics/DiagnosticDispatcher.php]: What about more loggers? I mean tracy panel could be taken as logger.

File: src/Tracy/templates/panel.phtml [hunk 4]
```diff
@@ -0,0 +1,85 @@
+<?php
```
Comment (f3l1x) [inline src/Tracy/templates/panel.phtml:1]: It looks like this file is weird formatted, don't you think? 

image

Comment (rattuscz) [reply src/Tracy/templates/panel.phtml:1]: yep, changed all to tabs.

File: src/DI/EventDispatcherExtension.php (hunk 1 above)
Comment (rattuscz) [reply src/DI/EventDispatcherExtension.php]: OK, i've realized it's not possible to use own listener anyway so it's always one of classes provided by this extension.

Comment (rattuscz) [reply src/DI/EventDispatcherExtension.php]: I don't know how to write the factory function for diagnostic version without creating second service for the wrapped dispatcher, as the LazyDispatcher requires container in constructor.
so is something like this ok?

```suggestion
		if ($config->lazy === true) {
			$factory = LazyEventDispatcher::class;
		} else {
			$factory = EventDispatcher::class;
		}

		if ($this->config->debug || $this->config->logger !== null) {
			$inner = $builder->addDefinition($this->prefix('innerDispatcher'))
				->setAutowired(false)
				->setFactory($factory);
			$eventDispatcherDefinition->setFactory(DiagnosticDispatcher::class, [$inner]);
		} else {
			$eventDispatcherDefinition->setFactory($factory);
		}
```

File: src/Diagnostics/DiagnosticDispatcher.php (hunk 3 above)
Comment (rattuscz) [reply src/Diagnostics/DiagnosticDispatcher.php]: reworked to multiple loggers with Tracy Panel as just a logger

File: src/Diagnostics/DiagnosticDispatcher.php (hunk 2 above)
Comment (rattuscz) [reply src/Diagnostics/DiagnosticDispatcher.php]: Panel was different from loggers, now simplified and this reference removed

**Comment by rattuscz** [comment]
ok I've rewritten it to use loggers only, but have one problem with php7.2 compatibility:

> Diagnostics/DiagnosticDispatcher.php : 77
> Method Symfony\Contracts\EventDispatcher\EventDispatcherInterface::dispatch() invoked with 2 parameters, 1 required.  

as the signature changed between symfony versions, how to make it work with phpstan? any ideas? :-)

**Comment by f3l1x** [comment]
I am ok with support only last versoin :-)

**Comment by petrparolek** [comment]
Great  work! I rebased this PR and dev-master and it works as I expected. Is it still in WIP?

**Comment by rattuscz** [comment]
Well i'm using this version on my project for over 2 years and had no troubles :-)
I was not able to finish fixing the stan/tests to get green results here.

**Comment by f3l1x** [comment]
Manually merged. Thank you for your work @rattuscz, well done. 👍 

https://github.com/contributte/event-dispatcher/releases/tag/v0.9.0

**Pull request closed without merging**